doc/warm-reboot/NSF_Manager_HLD.md

```diff
@@ -0,0 +1,280 @@
+
+
+#### Reconciliation Monitoring
+
+
+![alt_text](img/reconciliation.png)
+
+Upon bootup, components will detect warm-boot flag in STATE DB and reconcile to the state before reboot. Components that registered _reconciliation = true_ will update their warm-boot state = RECONCILED in STATE DB after they have completed reconciliation. NSF Manager will monitor the reconciliation status of these registered components during warm bootup. Subsequently, the components will start their normal operation and listen to new requests. The proposed reconciliation monitoring mechanism is similar to the existing mechanism except that components being monitored are based on the NSF details registration as opposed to hardcoded list of components.
+
+
+#### Component Warmboot States
+
+Based on the shutdown and reconciliation sequence mentioned above, a component will transition through the following states during warm reboot:
+
+
+
+*   Frozen (warm shutdown)
+*   Quiescent (warm shutdown)
+*   Checkpointed (optional, warm shutdown)
+*   Initialized (warm bootup)
+*   Reconciled (warm bootup)
+*   Failed (warm shutdown or bootup)
+
+
+
+![alt_text](img/warmboot-states.png)
+
+Upon receiving a freeze notification, a component will transition to _frozen_ state when it has stopped generating new intents for which it is the source. Subsequently, it will transition to _quiescent_ state when it has completed processing all requests. It might transition between _frozen_ and _quiescent_ states when it receives a new request from some other switch component and after it has completed processing all requests i.e. a component might transition between _frozen_ and _quiescent_ states if it processes an intent in freeze mode. If a component fails to perform its freeze routine i.e. fails to stop generating new intents for which it is the source then it will transition to _failed_ state. Upon receiving a checkpoint notification, a component will transition to _checkpointed_ state after it has completed checkpointing or to _failed_ state if checkpointing failed.
+
+After warm reboot, a component will transition to _initialized_ state after it has completed its initialization routine. It will transition to _reconciled_ state after it has completed reconciliation. It will transition to _failed_ state if its initialization or reconciliation fails.
+
+Components will update their state in STATE DB using [setWarmStartState()](https://github.com/sonic-net/sonic-swss-common/blob/master/common/warm_restart.cpp#L223) API during the different warm reboot stages. NSF Manager will monitor these NSF states in STATE DB to determine whether it needs to proceed with the next phase of the warm-boot orchestration or not.
+
+
+### SAI API
+
```

Comment:
Could add note that feature/new changes are independent of any SAI changes needed. There could be possible SAI changes needed for improvement.

Reply:
Added a note in SAI API section.